### ISSUE TITLE
twisted.names.test.test_dns.TestTCPController should override the messageReceived method.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6655]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kaizhang reported|
|-|-|
|Trac ID|trac#6655|
|Type|defect|
|Created|2013-07-30 08:33:26Z|
|Branch|https://github.com/twisted/twisted/tree/testtcpcontroller-messagereceived-6655|

`twisted.names.test.test_dns.TestTCPController` pretends to be a DNS query processor for a `twisted.names.dns.DNSProtocol`. It inherits from `twisted.names.test.test_dns.TestController`, a testing DNS query processor for a `twisted.names.dns.DNSDatagramProtocol`. However the `messageReceived` method for a `DNSProtocol` is different from the method for a `DNSDatagramProtocol`. There should be no `addr` parameter in the `messageReceived` method for a `DNSProtocol`. So `TestTCPController` should override the `messageReceived` method. 

<details><summary>Searchable metadata</summary>

```
trac-id__6655 6655
type__defect defect
reporter__kaizhang kaizhang
priority__normal normal
milestone__ 
branch__branches_testtcpcontroller_messagereceived_6655 branches/testtcpcontroller-messagereceived-6655
branch_author__kaizhang kaizhang
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1375173206000000 1375173206000000
changetime__1376382051000000 1376382051000000
version__None None
owner__kaizhang kaizhang

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|kaizhang commented|
|-|-|

(In [39378]) Branching to 'testtcpcontroller-messagereceived-6655'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/testtcpcontroller-messagereceived-6655)

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|kaizhang set owner to kaizhang|
|-|-|

Need to add news file.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|kaizhang removed owner|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/testtcpcontroller-messagereceived-6655)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj set owner to kaizhang|
|-|-|

Thanks Kai,

Notes:
 * merges cleanly
 * Tests pass on Fedora 19 x86_64
 * Builders all pass

Points:
 1. [branches/testtcpcontroller-messagereceived-6655/twisted/names/test/test_dns.py](branches/testtcpcontroller-messagereceived-6655/twisted/names/test/test_dns.py)
    1. Do you need to override messageReceived? I suggest you add an
       ```address=None``` argument to
       ```{TestController.messageReceived```. That would be consistent
       with twisted.names.server.DNSServerFactory.messageReceived.
       1. I did that locally and all the tests still pass.
    2. The name of the new test should end with "Messages" (plural) or
       make it something generic like
       "test_receiveQueryOrUnexpectedMessage".
       * It makes me think there should be more test coverage here to
         simulate queries and duplicate answers separately - but
         that's probably another ticket.
    3. Typo: test_receiveMessageNotInLiveMessage "in
       L{DNSDatagramProtocol.liveMessages} and
       L{DNSDatagramProtocol.resends}" should be "or
       L{DNSDatagramProtocol.resends}".
    4. "test_truncatedPacket" refers to self.controller rather than
       self.proto.controller. Maybe you should too, for consistency.
 2. [branches/testtcpcontroller-messagereceived-6655/twisted/names/topfiles/6655.bugfix](branches/testtcpcontroller-messagereceived-6655/twisted/names/topfiles/6655.bugfix)
    1. If you agree with point 1.1 above, then the newsfile either needs updating....OR
    2. ...change it to a .misc file. After all, this isn't a bug in
       any of the public APIs. I'd call it a case of missing test
       coverage.

Please merge after addressing or answering the numbered points above.

```

       Index: twisted/names/test/test_dns.py
       ===================================================================
       --- twisted/names/test/test_dns.py	(revision 39544)
       +++ twisted/names/test/test_dns.py	(working copy)
       @@ -712,7 +712,7 @@
                self.messages = []


       - def messageReceived(self, msg, proto, addr):
       - def messageReceived(self, msg, proto, addr=None):
             """
             Save the message so that it can be checked during the tests.
             """
       @@ -842,14 +842,7 @@
                self.connections.remove(proto)


       -    def messageReceived(self, msg, proto):
       -        """
       -        Save the message so that it can be checked during the tests.
       -        """
       -        self.messages.append((msg, proto))
       ```

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|kaizhang set status to `closed`|
|-|-|

(In [39577]) Merge testtcpcontroller-messagereceived-6655: Added addr=None argument to TestController.messageReceived

Author: kaizhang
Reviewer: rwall
Fixes: [#6655](https://github.com/twisted/twisted/issues/6655)

Added addr=None argument to twisted.names.test.test_dns.TestController.messageReceived. Now twisted.names.test.test_dns.TestTCPController.messageReceived can be used by twisted.names.dns.DNSProtocol.
